### PR TITLE
optimizer: refactor optimization flow, push filter condition to table. 

### PR DESCRIPTION
**Pull request by coocood**
This PR changed some of the optimization flow, prepare for implementing JOIN optimization.

1. Conditions are pushed to TableScan and IndexScan, so there is no need for Filter plan.
2. Choose best access method for a table in plan builder, the returned plan is the best plan, so we can remove Alternatives flow.
3. Estimate cost by computing filter rate which is based on expression type, so we don't need to evaluate expression before estimate cost.

### CODE REVIEW

File: optimizer/plan/filterrate.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+		return computeIsNullFilterRate(x)
+	case *ast.IsTruthExpr:
+		return computeIsTrueFilterRate(x)
+	case *ast.ParenthesesExpr:
+		return computeFilterRate(x.Expr)
+	case *ast.PatternInExpr:
+		return computePatternInFilterRate(x)
+	case *ast.PatternLikeExpr:
+		return computePatternLikeFilterRate(x)
+	case *ast.ColumnNameExpr:
```
Comment (qiuyesuifeng) [inline optimizer/plan/filterrate.go]: Please keep in order.

File: optimizer/plan/filterrate.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	rateBetween       = 0.3
+	rateGreaterOrLess = 0.4
+	rateIsFalse       = 0.01
+	rateIsNull        = 0.01
+	rateLike          = 0.1
+)
+
+// computeFilterRate computes the filter rate for an expression.
+// It only depends on the expression type, not the expression value.
+// The expr parameter should contains only one column name.
```
Comment (qiuyesuifeng) [inline optimizer/plan/filterrate.go]: should contain

File: optimizer/plan/filterrate.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	case *ast.ParenthesesExpr:
+		return computeFilterRate(x.Expr)
+	case *ast.PatternInExpr:
+		return computePatternInFilterRate(x)
+	case *ast.PatternLikeExpr:
+		return computePatternLikeFilterRate(x)
+	case *ast.ColumnNameExpr:
+		return rateFull
+	}
+	return rateFull
```
Comment (qiuyesuifeng) [inline optimizer/plan/filterrate.go]: How about using fullRate?

Comment (qiuyesuifeng) [inline optimizer/plan/filterrate.go]: Ignore it, seems the same.

File: optimizer/plan/planbuilder.go [hunk 4]
```diff
@@ -202,38 +192,90 @@ func (b *planBuilder) buildJoin(from *ast.Join) Plan {
...
-			conditions = append(conditions, x.L)
-			conditions = append(conditions, b.splitWhere(x.R)...)
+func (b *planBuilder) buildAllAccessMethodsPlan(tn *ast.TableName, conditions []ast.ExprNode) []Plan {
+	var candidates []Plan
+	p := &TableScan{
+		Table: tn.TableInfo,
+	}
+	p.SetFields(tn.GetResultFields())
+	var pkName model.CIStr
+	if p.Table.PKIsHandle {
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder.go]: How about extracting this to a function?

File: optimizer/plan/planbuilder.go [hunk 5]
```diff
@@ -355,3 +397,78 @@ func buildResultField(tableName, name string, tp byte, size int) *ast.ResultFiel
 		Expr:         expr,
 	}
 }
+
+// matchOrder checks if the plan has the same ordering as items.
+func matchOrder(p Plan, items []*ast.ByItem) bool {
+	switch x := p.(type) {
+	case *TableScan:
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder.go]: Please keep in order.

File: optimizer/plan/planbuilder.go [hunk 6]
```diff
@@ -355,3 +397,78 @@ func buildResultField(tableName, name string, tp byte, size int) *ast.ResultFiel
...
+				return false
+			}
+			var rf *ast.ResultField
+			switch y := item.Expr.(type) {
+			case *ast.ColumnNameExpr:
+				rf = y.Refer
+			case *ast.PositionExpr:
+				rf = y.Refer
+			}
+			if rf == nil {
```
Comment (qiuyesuifeng) [inline optimizer/plan/planbuilder.go]: Please unify it.

File: optimizer/plan/filterrate.go (hunk 1 above)
Comment (shenli) [inline optimizer/plan/filterrate.go]: We do not need this case.

File: optimizer/plan/filterrate.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		}
+		return rate
+	case opcode.EQ:
+		return rateEqual
+	case opcode.GT, opcode.GE, opcode.LT, opcode.LE:
+		return rateGreaterOrLess
+	case opcode.NE:
+		return rateNotEqual
+	}
+	return 1
```
Comment (shenli) [inline optimizer/plan/filterrate.go]: Use a const for 1.

File: optimizer/plan/filterrate.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+		return computePatternInFilterRate(x)
+	case *ast.PatternLikeExpr:
+		return computePatternLikeFilterRate(x)
+	case *ast.ColumnNameExpr:
+		return rateFull
+	}
+	return rateFull
+}
+
+func computeBinopFilterRate(expr *ast.BinaryOperationExpr) float64 {
```
Comment (shenli) [inline optimizer/plan/filterrate.go]: Need a document or more comments to describe the algorithm.

File: executor/builder.go [hunk 9]
```diff
@@ -79,13 +77,22 @@ func (b *executorBuilder) build(p plan.Plan) Executor {
...
 	table, _ := b.is.TableByID(v.Table.ID)
-	return &TableScanExec{
+	e := &TableScanExec{
 		t:          table,
 		fields:     v.Fields(),
 		ctx:        b.ctx,
 		ranges:     v.Ranges,
 		seekHandle: math.MinInt64,
 	}
+	if len(v.FilterConditions) != 0 {
```
Comment (shenli) [inline executor/builder.go]: Extract to a function?

**Comment by shenli** [comment]
LGTM

File: optimizer/plan/cost.go [hunk 10]
```diff
@@ -117,33 +87,9 @@ func (c *costEstimator) tableScan(v *TableScan) {
...
-			rowCount += MiddleRangeCount
-		}
-		// If the index has too many ranges, the row count may exceed the default row count.
-		// Make sure the cost is lower than full range.
-		if rowCount >= FullRangeCount {
-			rowCount = FullRangeCount - 1
-		}
+	var rowCount float64 = FullRangeCount
+	for _, con := range v.AccessConditions {
+		rowCount *= guesstimateFilterRate(con)
```
Comment (qiuyesuifeng) [inline optimizer/plan/cost.go]: Why not using estimate?

Comment (coocood) [inline optimizer/plan/cost.go]: MySQL uses `guesstimate` for this kind of rule.
Estimate is based on index statistics.

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**